python/triton/compiler/code_generator.py

```diff
@@ -1192,11 +1226,14 @@ def visit_Assert(self, node) -> Any:
         return language.core.device_assert(test, msg, _semantic=self.semantic)
 
     def call_JitFunction(self, fn: JITFunction, args, kwargs):
-        args = inspect.getcallargs(fn.fn, *args, **kwargs)
-        args = [args[name] for name in fn.arg_names]
+        call_args = inspect.getcallargs(fn.fn, *args, **kwargs)
+        args = [call_args[name] for name in fn.arg_names]
         for i, arg in enumerate(args):
             if isinstance(arg, (language.dtype, float, int, bool, JITFunction)):
                 args[i] = language.core.constexpr(arg)
+        args_mut = [[name, None]
+                    for (name, arg) in zip(fn.arg_names, args)
+                    if arg is not None and getattr(arg.type, "__triton_mutable__", False)]
```

Comment:
What if `__triton_mutable__` isn't on the top level type, e.g. `tuple(aggregate)`?

Reply:
I thought about banning putting mutable values inside non-mutable ones, but in fact putting any aggregate inside another will take a copy because they aren't actually reference semantic.